GPU/GPUTracking/DataCompression/GPUTPCCompressionTrackModel.cxx

```diff
@@ -112,4 +112,923 @@ GPUd() int GPUTPCCompressionTrackModel::Mirror()
+
+  float dL = std::copysignf(dS * mTrk.dlds, -1.f); // we are in flight direction
+
+  float& mC40 = mC[10];
+  float& mC41 = mC[11];
+  float& mC42 = mC[12];
+  float& mC43 = mC[13];
+  float& mC44 = mC[14];
+
+  float dLmask = 0.f;
+  bool maskMS = (CAMath::Abs(dL) < mMaterial.DLMax);
+  if (maskMS) {
+    dLmask = dL;
+  }
+  float dLabs = CAMath::Abs(dLmask);
+  float corr = 1.f - mMaterial.EP2 * dLmask;
+
+  float corrInv = 1.f / corr;
+  mTrk.px *= corrInv;
+  mTrk.py *= corrInv;
+  mTrk.pz *= corrInv;
+  mTrk.pt *= corrInv;
+  mTrk.p *= corrInv;
+  mTrk.qpt *= corr;
+
+  mP[4] *= corr;
+
+  mC40 *= corr;
+  mC41 *= corr;
+  mC42 *= corr;
+  mC43 *= corr;
+  mC44 = mC44 * corr * corr + dLabs * mMaterial.sigmadE2;
+
+  return 0;
+}
+
+GPUd() void GPUTPCCompressionTrackModel::getBxByBz(float cosAlpha, float sinAlpha, float x, float y, float z, float b[3]) const
+{
+  float xGlb = x * cosAlpha - y * sinAlpha;
+  float yGlb = x * sinAlpha + y * cosAlpha;
+  float bb[3];
+  mField->GetField(xGlb, yGlb, z, bb);
+  // rotate field to local coordinates
+  b[0] = bb[0] * cosAlpha + bb[1] * sinAlpha;
+  b[1] = -bb[0] * sinAlpha + bb[1] * cosAlpha;
+  b[2] = bb[2];
+}
+
+GPUd() float GPUTPCCompressionTrackModel::getBz(float x, float y, float z) const
+{
+  float xGlb = x * mCosAlpha - y * mSinAlpha;
+  float yGlb = x * mSinAlpha + y * mCosAlpha;
+  return mField->GetFieldBz(xGlb, yGlb, z);
+}
+
+GPUd() void GPUTPCCompressionTrackModel::updatePhysicalTrackValues(PhysicalTrackModel& trk)
+{
+  float px = trk.px;
+  if (CAMath::Abs(px) < 1.e-4f) {
+    px = copysign(1.e-4f, px);
+  }
+
+  trk.pt = sqrt(px * px + trk.py * trk.py);
+  float pti = 1.f / trk.pt;
+  trk.p = sqrt(px * px + trk.py * trk.py + trk.pz * trk.pz);
+  trk.sinphi = trk.py * pti;
+  trk.cosphi = px * pti;
+  trk.secphi = trk.pt / px;
+  trk.dzds = trk.pz * pti;
+  trk.dlds = trk.p * pti;
+  trk.qpt = trk.q * pti;
+}
+
+GPUd() void GPUTPCCompressionTrackModel::changeDirection()
+{
+  mTrk.py = -mTrk.py;
+  mTrk.pz = -mTrk.pz;
+  mTrk.q = -mTrk.q;
+  mTrk.sinphi = -mTrk.sinphi;
+  mTrk.dzds = -mTrk.dzds;
+  mTrk.qpt = -mTrk.qpt;
+  updatePhysicalTrackValues(mTrk);
+
+  mC[3] = -mC[3];
+  mC[4] = -mC[4];
+  mC[6] = -mC[6];
+  mC[7] = -mC[7];
+  mC[10] = -mC[10];
+  mC[11] = -mC[11];
+}
+
+GPUd() int GPUTPCCompressionTrackModel::rotateToAlpha(float newAlpha)
+{
+  //
+  // Rotate the track coordinate system in XY to the angle newAlpha
+  // return value is error code (0==no error)
+  //
+
+  float newCosAlpha = CAMath::Cos(newAlpha);
+  float newSinAlpha = CAMath::Sin(newAlpha);
+
+  float cc = newCosAlpha * mCosAlpha + newSinAlpha * mSinAlpha; // cos(newAlpha - mAlpha);
+  float ss = newSinAlpha * mCosAlpha - newCosAlpha * mSinAlpha; // sin(newAlpha - mAlpha);
+
+  PhysicalTrackModel t0 = mTrk;
+
+  float x0 = mTrk.x;
+  float y0 = mTrk.y;
+  float px0 = mTrk.px;
+  float py0 = mTrk.py;
+
+  if (CAMath::Abs(mP[2]) >= GPUCA_MAX_SIN_PHI || CAMath::Abs(px0) < (1 - GPUCA_MAX_SIN_PHI)) {
```

Comment:
Please replace GPUCA_MAX_SIN_PHI by a constexpr value inside the header, it should be self-contained and not use any external parameters.

Reply:
Done